File: apps/api/src/callcraft_engine/schema.py
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class PlatformDataType(str, Enum):
    STRING = "string"
    TEXT = "text"
    INTEGER = "integer"
    NUMBER = "number"
    BOOLEAN = "boolean"
    EMAIL = "email"
    PHONE = "phone"
    DATE = "date"
    DATETIME = "datetime"
    CURRENCY = "currency"
    ENUM = "enum"
    OBJECT = "object"
    ARRAY = "array"


class FieldDefinition(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    type: PlatformDataType
    required: bool = True
    description: Optional[str] = None
    default: Optional[Any] = None
    enum_values: Optional[List[str]] = Field(default=None, alias="values")
    properties: Optional[Dict[str, "FieldDefinition"]] = None
    items: Optional["FieldDefinition"] = None


FieldDefinition.model_rebuild()


class ResponseSchema(BaseModel):
    title: Optional[str] = None
    properties: Dict[str, FieldDefinition]

File: crates/ocr-engine/src/schema.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DataType {
    String,
    Text,
    Integer,
    Number,
    Boolean,
    Email,
    Phone,
    Date,
    DateTime,
    Currency,
    Enum { values: Vec<String> },
    Object { properties: HashMap<String, FieldDefinition> },
    Array { items: Box<FieldDefinition> },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FieldDefinition {
    pub r#type: DataType,
    #[serde(default)]
    pub required: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResponseSchema {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub properties: HashMap<String, FieldDefinition>,
}

impl ResponseSchema {
    pub fn new(properties: HashMap<String, FieldDefinition>) -> Self {
        Self {
            title: None,
            properties,
        }
    }
}
